pkg/bgp/speaker/events.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package speaker
+
+import (
+	"go.universe.tf/metallb/pkg/k8s/types"
+	metallbspr "go.universe.tf/metallb/pkg/speaker"
+)
+
+type svcEvent struct {
+	name string
+	svc  *metallbspr.Service
+	eps  *metallbspr.Endpoints
+}
+type epEvent svcEvent
+
+// nodeEvent must be a pointer to a map because maps are not hashable, aka
+// cannot be keys in a map, which queue is.
+type nodeEvent *map[string]string
+
+// run runs the reconciliation loop, fetching events off of the queue to
+// process. The events supported are svcEvent, epEvent, and nodeEvent. This
+// loop is only stopped (implicitly) when the Agent is shutting down.
+//
+// Adapted from go.universe.tf/metallb/pkg/k8s/k8s.go.
+func (s *Speaker) run() {
+	for {
+		key, quit := s.queue.Get()
+		if quit {
+			return
+		}
+		st := s.do(key)
+		switch st {
+		case types.SyncStateError:
+			s.queue.AddRateLimited(key)
```

Comment:
Do we want to have rate limiting in service events? What will happen if we receive 1k service events and this watcher can't keep up with all events? a) Does it mean all other watchers will be blocked by this watcher? b) Or we have an independent queue that will store all svc events received and process them in parallel? If it is b) what guarantees do we have that we won't have concurrent issues while handling events for a svc that no longer exists, or it was deleted and re-created with the same name?

Reply:
Good point. The way it's implemented is (b). The queue processing runs in a goroutine and all the handlers do is push into the queue, which is not blocking. So other handlers should not be affected. However, you also make a good 2nd point about concurrent issues. Rather than thinking about a more complex way to handle them, let's use a simple non-rate limiting queue. Plus, MetalLB isn't performing any complex operations; it's simply allocating an IP. Failing to do that is the only real consequential failure. I don't suspect that to happen often, besides the case where the user exhausts their IP pool . Therefore, we don't really need to rate-limit the retries.